gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/FSPathAlterationFlowGraphListener.java

```diff
@@ -0,0 +1,110 @@
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.gobblin.runtime.api.TopologySpec;
+import org.apache.gobblin.service.modules.template_catalog.FSFlowTemplateCatalog;
+import org.apache.gobblin.util.filesystem.PathAlterationListener;
+import org.apache.gobblin.util.filesystem.PathAlterationObserver;
+import org.apache.hadoop.fs.Path;
+
+@Slf4j
+public class FSPathAlterationFlowGraphListener extends BaseFlowGraphListener implements PathAlterationListener {
+
+  private File graphDir;
+  CountDownLatch initComplete;
+
+  public FSPathAlterationFlowGraphListener(Optional<? extends FSFlowTemplateCatalog> flowTemplateCatalog,
+      FlowGraph graph, Map<URI, TopologySpec> topologySpecMap, String baseDirectory, String flowGraphFolderName,
+      String javaPropsExtentions, String hoconFileExtensions, CountDownLatch initComplete) {
+    super(flowTemplateCatalog, graph, topologySpecMap, baseDirectory, flowGraphFolderName, javaPropsExtentions, hoconFileExtensions);
+    this.graphDir = new File(baseDirectory);
+    this.initComplete = initComplete;
+    // Populate the flowgraph with any existing files
+    if (!this.graphDir.exists()) {
+      throw new RuntimeException(String.format("Flowgraph directory at path %s does not exist!", graphDir));
+    }
+    this.populateFlowGraphAtomically();
+  }
+
+  public void onStart(final PathAlterationObserver observer) {
+  }
+
+  public void onFileCreate(final Path path) {
+  }
+
+  public void onFileChange(final Path path) {
+  }
+
+  public void onStop(final PathAlterationObserver observer) {
+  }
+
+  public void onDirectoryCreate(final Path directory) {
+  }
+
+  public void onDirectoryChange(final Path directory) {
+  }
+
+  public void onDirectoryDelete(final Path directory) {
+  }
+
+  public void onFileDelete(final Path path) {
+  }
+
+  @Override
+  public void onCheckDetectedChange() {
+    log.info("Detecting change in flowgraph files, reloading flowgraph");
+    this.populateFlowGraphAtomically();
+  }
+
+  private void populateFlowGraphAtomically() {
+    FlowGraph newFlowGraph = new BaseFlowGraph();
+    try {
+      List<Path> edges = new ArrayList<>();
+      // All nodes must be added first before edges, otherwise edges may have a missing source or destination.
+      Files.walk(this.graphDir.toPath()).forEach(fileName -> {
+        if (!Files.isDirectory(fileName)) {
+          if (checkFileLevelRelativeToRoot(new Path(fileName.toString()), NODE_FILE_DEPTH)) {
+            addDataNode(newFlowGraph, fileName.toString());
+          } else if (checkFileLevelRelativeToRoot(new Path(fileName.toString()), EDGE_FILE_DEPTH)) {
+            edges.add(new Path(fileName.toString()));
+          }
+        }
+      });
+      for (Path edge: edges) {
+        addFlowEdge(newFlowGraph, edge.toString());
```

Comment:
Unfortunately it's due to the implementation of the file listener, it was originally written for Gobblin pipelines and thus it uses the Hadoop Path references. There are a lot of usages of Hadoop Path objects in GaaS that I find unnecessary, especially because doing so requires GaaS to stay on outdated versions of Java. Hence I try to avoid having new classes rely on the Hadoop Path library but for compatibility it seems hard to avoid.